crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,618 @@
+            } else {
+                Vec::new()
+            },
+            intern_context: InternContext::new(),
+            qubits_native: Vec::new(),
+            clbits_native: Vec::new(),
+            qubit_indices_native: HashMap::new(),
+            clbit_indices_native: HashMap::new(),
+            qubits: PyList::empty(py).into_py(py),
+            clbits: PyList::empty(py).into_py(py),
+        };
+        if let Some(qubits) = qubits {
+            for bit in qubits.iter()? {
+                self_.add_qubit(py, bit?)?;
+            }
+        }
+        if let Some(clbits) = clbits {
+            for bit in clbits.iter()? {
+                self_.add_clbit(py, bit?)?;
+            }
+        }
+        if let Some(data) = data {
+            self_.extend(py, data)?;
+            self_.reserve(py, reserve);
+        }
+        Ok(self_)
+    }
+
+    pub fn __reduce__(self_: &PyCell<CircuitData>, py: Python<'_>) -> PyResult<PyObject> {
+        let ty: &PyType = self_.get_type();
+        let args = {
+            let self_ = self_.borrow();
+            (
+                self_.qubits.clone_ref(py),
+                self_.clbits.clone_ref(py),
+                None::<()>,
+                self_.data.len(),
+            )
+        };
+        Ok((ty, args, None::<()>, self_.iter()?).into_py(py))
+    }
+
+    /// Registers a :class:`.Qubit` instance.
+    ///
+    /// Args:
+    ///     bit (:class:`.Qubit`): The qubit to register.
+    pub fn add_qubit(&mut self, py: Python<'_>, bit: &PyAny) -> PyResult<()> {
+        let idx = self.qubits_native.len() as u32;
+        self.qubit_indices_native.insert(BitAsKey::new(bit)?, idx);
+        self.qubits_native.push(bit.into_py(py));
+        self.qubits.as_ref(py).append(bit)
+    }
+
+    /// Registers a :class:`.Clbit` instance.
+    ///
+    /// Args:
+    ///     bit (:class:`.Clbit`): The clbit to register.
+    pub fn add_clbit(&mut self, py: Python<'_>, bit: &PyAny) -> PyResult<()> {
+        let idx = self.clbits_native.len() as u32;
+        self.clbit_indices_native.insert(BitAsKey::new(bit)?, idx);
+        self.clbits_native.push(bit.into_py(py));
+        self.clbits.as_ref(py).append(bit)
+    }
+
+    /// Performs a shallow copy.
+    ///
+    /// Returns:
+    ///     CircuitData: The shallow copy.
+    pub fn copy(&self, py: Python<'_>) -> PyResult<Self> {
+        Ok(CircuitData {
+            data: self.data.clone(),
+            // TODO: reuse intern context once concurrency is properly
+            //  handled.
+            intern_context: self.intern_context.clone(),
+            qubits_native: self.qubits_native.clone(),
+            clbits_native: self.clbits_native.clone(),
+            qubit_indices_native: self.qubit_indices_native.clone(),
+            clbit_indices_native: self.clbit_indices_native.clone(),
+            qubits: self.qubits.clone_ref(py),
+            clbits: self.clbits.clone_ref(py),
+        })
+    }
```

Comment:
Heh, I actually didn't realize that after a few iterations, all of these fields ended up being simple calls to `clone`.

I'd like to leave this as is however since in a future PR, the logic to share `self.intern_context` will not just be a clone.